src/3d/qgslayoutitem3dmap.cpp

```diff
@@ -41,28 +44,149 @@ int QgsLayoutItem3DMap::type() const
+    painter->restore();
+    return;
+  }
+
+  if ( !mCapturedImage.isNull() )
+  {
+    painter->drawImage( r, mCapturedImage );
+    painter->restore();
+    return;
+  }
+
+  // we do not have a cached image of the rendered scene - let's request one from the engine
+
+  if ( mLayout->renderContext().isPreviewRender() )
+  {
+    painter->drawText( r, Qt::AlignCenter, tr( "Loading" ) );
+    painter->restore();
+  }
+
   QSizeF sizePixels = mLayout->renderContext().measurementConverter().convert( sizeWithUnits(), QgsUnitTypes::LayoutPixels ).toQSizeF();
-  engine.setSize( QSize( static_cast<int>( std::ceil( sizePixels.width() ) ),
-                         static_cast<int>( std::ceil( sizePixels.height() ) ) ) );
+  QSize sizePixelsInt = QSize( static_cast<int>( std::ceil( sizePixels.width() ) ),
+                               static_cast<int>( std::ceil( sizePixels.height() ) ) );
 
-  Qgs3DMapScene *scene = new Qgs3DMapScene( *mSettings, &engine );
-  engine.setRootEntity( scene );
+  if ( !mEngine )
+  {
+    mEngine.reset( new QgsOffscreen3DEngine );
+    connect( mEngine.get(), &QgsAbstract3DEngine::imageCaptured, this, &QgsLayoutItem3DMap::onImageCaptured );
 
-  scene->cameraController()->setCameraPose( mCameraPose );
+    mEngine->setSize( sizePixelsInt );
 
-  // XXX this should not be needed, but without it the scene often
-  // is not completely ready (e.g. a missing terrain tile).
-  // leaving it here until a more robust solution is found...
-  Qgs3DUtils::captureSceneImage( engine, scene );
+    mScene = new Qgs3DMapScene( *mSettings, mEngine.get() );
+    connect( mScene, &Qgs3DMapScene::sceneStateChanged, this, &QgsLayoutItem3DMap::onSceneStateChanged );
 
-  QImage img = Qgs3DUtils::captureSceneImage( engine, scene );
+    mEngine->setRootEntity( mScene );
+  }
 
-  QgsRenderContext &ctx = context.renderContext();
-  ctx.painter()->drawImage( 0, 0, img );
+  if ( mEngine->size() != sizePixelsInt )
+    mEngine->setSize( sizePixelsInt );
+
+  mScene->cameraController()->setCameraPose( mCameraPose );
+
+  if ( mLayout->renderContext().isPreviewRender() )
+  {
+    onSceneStateChanged();
+  }
+  else
+  {
+    // we can't just request a capture and hope it will arrive at some point later.
+    // this is not a preview, we need the rendered scene now!
+    if ( mDrawing )
+      return;
+    mDrawing = true;
+    Qgs3DUtils::captureSceneImage( *mEngine.get(), mScene );
+    QImage img = Qgs3DUtils::captureSceneImage( *mEngine.get(), mScene );
+    painter->drawImage( r, img );
+    painter->restore();
+    mDrawing = false;
+  }
+}
+
+void QgsLayoutItem3DMap::onImageCaptured( const QImage &img )
+{
+  mCapturedImage = img;
+  update();
+}
+
+void QgsLayoutItem3DMap::onSceneStateChanged()
+{
+  if ( mCapturedImage.isNull() && mScene->sceneState() == Qgs3DMapScene::Ready )
+  {
+    mEngine->requestCaptureImage();
+  }
+}
+
+void QgsLayoutItem3DMap::onSizePositionChanged()
+{
+  mCapturedImage = QImage();
```

Comment:
You'll also need to implement QgsLayoutItem::refresh and invalidate the cache there. That will allow manual user layout refreshes to force a new render too.

Reply:
added